multi_partner_learning.py

```diff
@@ -0,0 +1,532 @@
+            self.models_weights_list[partner_index] = model_for_round.get_weights()
+
+            # Update iterative results
+            self.update_iterative_results(partner_index, history)
+
+        logger.debug("End of seqavg collaborative round.")
+
+    def split_in_minibatches(self):
+        """Split the dataset passed as argument in mini-batches"""
+
+        # Create the indices where to split
+        split_indices = np.arange(1, self.minibatch_count + 1) / self.minibatch_count
+
+        # Iterate over all partners and split their datasets
+        for partner_index, partner in enumerate(self.partners_list):
+            # Shuffle the dataset
+            idx = np.random.permutation(len(partner.x_train))
+            x_train, y_train = partner.x_train[idx], partner.y_train[idx]
+
+            # Split the samples and labels
+            self.minibatched_x_train[partner_index] = np.split(x_train, (split_indices[:-1] * len(x_train)).astype(int))
+            self.minibatched_y_train[partner_index] = np.split(y_train, (split_indices[:-1] * len(y_train)).astype(int))
+
+    def prepare_aggregation_weights(self, input_weights):
+        """Returns a list of weights for the weighted average aggregation of model weights"""
+
+        if self.aggregation_weighting == "uniform":
+            self.aggregation_weights = [1 / self.partners_count] * self.partners_count
+        elif self.aggregation_weighting == "data_volume":
+            partners_sizes = [len(partner.x_train) for partner in self.partners_list]
+            self.aggregation_weights = partners_sizes / np.sum(partners_sizes)
+        elif self.aggregation_weighting == "local_score":
+            self.aggregation_weights = input_weights / np.sum(input_weights)
+        else:
+            raise NameError("The aggregation_weighting value [" + self.aggregation_weighting + "] is not recognized.")
+
+    def aggregate_model_weights(self):
+        """Aggregate model weights from the list of models, with a weighted average"""
+
+        weights_per_layer = list(zip(*self.models_weights_list))
+        new_weights = list()
+
+        for weights_for_layer in weights_per_layer:
+            avg_weights_for_layer = np.average(
+                np.array(weights_for_layer), axis=0, weights=self.aggregation_weights
+            )
+            new_weights.append(list(avg_weights_for_layer))
+
+        return new_weights
+
+    @staticmethod
+    def build_aggregated_model(new_weights):
+        """Generate a new model initialized with weights passed as arguments"""
+
+        aggregated_model = utils.generate_new_cnn_model()
+        aggregated_model.set_weights(new_weights)
+        aggregated_model.compile(
+            loss=keras.losses.categorical_crossentropy,
+            optimizer="adam",
+            metrics=["accuracy"],
+        )
+
+        return aggregated_model
+
+    @staticmethod
+    def build_from_previous_model(previous_model):
```

Comment:
Where is this used ?

Reply:
This is a kind of "model deep copy" function. I use it in the seq and seqavg.

Reply:
Which line ? I could not find where it was called

Reply:
Ah yeah you are right I didn't need it anymore once I started storing weights instead of models. You might need it though when you'll address the `clear_session()` #142 issue (potentially to be transformed to a `build_from_weights` function).

Reply:
Ahhh, in fact this is exactly the role of the function you commented above: `build_aggregated_model()` to be renamed to `build_from_weights()`. Ok so I'll simplify these